### PR TITLE
Fix Syntax Highlighting

### PR DESCRIPTION
**Pull request by takanuva15**
Hey, I know it's been some time but I finally got around to exploring the bnf/jflex code to try and fix it. 

Basically I removed all the classes that were causing errors and I commented out the original bnf/lex code that was there. In its place, I put a simplified version of the bnf that _only_ checks for line comments currently, and it should give 0 errors for any script that includes line comments (I had to check the rules for line comments in AutoHotkey; the only thing missing here is support for comments in multiline strings which will be fixed later when string bnf is added). I also added some junit tests and modified the github actions .yml file based on your older PR which will run all the tests to verify. 

I also fixed the version compatibility issues with IDEA 2020, so you should be able to publish a new version of the plugin after this PR (v0.4.0). (I was forced to drop support for IDEA 2019 because JetBrains changed some of the APIs significantly). In the next PR, I will add support for block comments and a GitHub action to check the test coverage.

### CODE REVIEW

**Comment by takanuva15** [comment]
Hmm the build looks stuck since the github yaml file was updated. Seems this has been reported before; something might need to be updated in the repository settings. See here as example: https://github.community/t/github-actions-pull-requests-always-include-some-outdated-checks/16157/5

**Comment by MTRNord** [comment]
> 
> 
> Hmm the build looks stuck since the github yaml file was updated. Looks like the build checks or something might need to be updated in the repository branch settings. (Seems this has been reported before, see [here](https://github.community/t/github-actions-pull-requests-always-include-some-outdated-checks/16157/5) for example: )
> 
> image

Fixed :)

**Comment by MTRNord** [comment]
I will review this later today. Also sorry that I did not work on this for a long time. Just didnt get to it since University started,

**Comment by takanuva15** [comment]
> I will review this later today. Also sorry that I did not work on this for a long time. Just didnt get to it since University started,

Lol yea, university work always swallowed most of any free time I had while I was at Uni.

File: src/main/java/de/nordgedanken/auto_hotkey/ide/highlighter/AhkColors.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
+package de.nordgedanken.auto_hotkey.ide.highlighter
+
+import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
+import com.intellij.openapi.editor.colors.TextAttributesKey
+import com.intellij.openapi.options.colors.AttributesDescriptor
+
+enum class AhkColor(humanName: String, default: TextAttributesKey? = null) {
+    LINE_COMMENT("Comments//Line Comments", DefaultLanguageHighlighterColors.LINE_COMMENT),
+//    FUNCTION("Functions//Function declaration", DefaultLanguageHighlighterColors.FUNCTION_DECLARATION),
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/ide/highlighter/AhkColors.kt:9]: Just remove the commented lines. Thats the beauty of version control. No need to keep them. We can always look at old commits

File: src/main/java/de/nordgedanken/auto_hotkey/ide/highlighter/AhkHighlightAnnotator.kt [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.Annotator
+import com.intellij.lang.annotation.HighlightSeverity.INFORMATION
+import com.intellij.psi.PsiElement
+//import de.nordgedanken.auto_hotkey.lang.psi.AhkHotkeyAssignment
+
+class AhkHighlightAnnotator : Annotator {
+    override fun annotate(psiElem: PsiElement, holder: AnnotationHolder) {
+//        if (psiElem is AhkHotkeyAssignment) {
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/ide/highlighter/AhkHighlightAnnotator.kt:11]: Same as above

File: src/main/java/de/nordgedanken/auto_hotkey/ide/highlighter/AhkSyntaxHighlighter.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+class AhkSyntaxHighlighter : SyntaxHighlighterBase() {
+    override fun getHighlightingLexer(): Lexer = AhkLexerAdapter()
+
+    override fun getTokenHighlights(tokenType: IElementType): Array<TextAttributesKey> =
+            pack(map(tokenType)?.textAttributesKey)
+
+    companion object {
+        fun map(tokenType: IElementType): AhkColor? = when (tokenType) {
+            LINE_COMMENT -> AhkColor.LINE_COMMENT
+//            HOTKEY_ASSIGNMENT -> AhkColor.FUNCTION
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/ide/highlighter/AhkSyntaxHighlighter.kt:21]: Same as above

File: src/main/java/de/nordgedanken/auto_hotkey/lang/lexer/AutoHotkey.flex [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+//generic
+CHAR_SPECIAL=[^\s[:letter:][:digit:]]
+TEXT=\w+
+WHITESPACE_HOZ=\p{Blank}+
+CRLF=\R
+
+LINE_COMMENT=;.*
+
+//language-specific
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/lang/lexer/AutoHotkey.flex:28]: Same as above

File: src/main/java/de/nordgedanken/auto_hotkey/lang/lexer/AutoHotkey.flex [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+	{LINE_COMMENT}      { return LINE_COMMENT; }
+	[^]                 {
+							yypushback(1);      // cancel parsed char (no comment typed here)
+							yybegin(YYINITIAL); // and try to parse it again in <YYINITIAL>
+						}
+}
+
+[^] { return BAD_CHARACTER; }
+
+/*
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/lang/lexer/AutoHotkey.flex:58]: Same as above

File: src/main/java/de/nordgedanken/auto_hotkey/lang/parser/AhkParserDefinition.kt [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    override fun createParser(project: Project?): PsiParser = AhkParser()
+
+    override fun getFileNodeType() = IFileElementType(AhkLanguage)
+
+    override fun createFile(viewProvider: FileViewProvider): PsiFile = AhkFile(viewProvider)
+
+    override fun spaceExistanceTypeBetweenTokens(left: ASTNode, right: ASTNode) = ParserDefinition.SpaceRequirements.MAY
+
+/*    override fun spaceExistanceTypeBetweenTokens(left: ASTNode, right: ASTNode): ParserDefinition.SpaceRequirements {
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/lang/parser/AhkParserDefinition.kt:35]: Same as above

File: src/main/java/de/nordgedanken/auto_hotkey/lang/parser/AutoHotkey.bnf [hunk 7]
```diff
@@ -1,76 +1,68 @@
...
+AhkFile ::= Item*
+private Item ::= Line | LINE_COMMENT | CRLF //Note that only CRLF can follow a line comment
 
-  consumeTokenMethod("((?!.*_with_recover).*_recover)|(.*Expr)") = "consumeTokenFast"
-}
+Line ::= (TEXT | CHAR_SPECIAL)+ (WHITESPACE_HOZ LINE_COMMENT?)? //optionally match a line comment if there's whitespace while parsing the line
+
+
+
+//Hotkey ::= ('#'|'!'|'^'|'+'|'<'|'>'|'<^>!'|'*'|'~'|'$')* (id | CHAR_SPECIAL) (WHITESPACE '&' WHITESPACE (id | CHAR_SPECIAL))? (WHITESPACE 'Up')? { rightAssociative=true}
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/lang/parser/AutoHotkey.bnf:52]: Same as above

File: src/main/java/de/nordgedanken/auto_hotkey/lang/psi/AhkTokenType.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
+package de.nordgedanken.auto_hotkey.lang.psi
+
+import com.intellij.psi.tree.IElementType
+import de.nordgedanken.auto_hotkey.lang.core.AhkLanguage
+
+//import de.nordgedanken.auto_hotkey.psi.AHKTypes.*
+
+open class AhkTokenType(debugName: String) : IElementType(debugName, AhkLanguage)
+
+//val AHK_ITEMS = tokenSetOf(
```
Comment (MTRNord) [inline src/main/java/de/nordgedanken/auto_hotkey/lang/psi/AhkTokenType.kt:10]: Same as above

**Review by MTRNord — CHANGES_REQUESTED** [review]
Looks fine except the commented stuff which just can be removed. We always are able to go back in the git history

**Review by MTRNord — APPROVED** [review]
Looks fine :) Feel free to press merge if you are ready to merge this. (Not sure if this is WIP or ready? :D)

**Comment by MTRNord** [comment]
Oh I apparently went really strict with checks... Lets loosen that up :3

**Comment by takanuva15** [comment]
> Looks fine :) Feel free to press merge if you are ready to merge this. (Not sure if this is WIP or ready? :D)

I don't see any 'Merge' button... I think only you have write access to this repository so you would have to merge the PR. 
(This is ready to merge; I would just make a 'draft pull request' or add a comment if I wasn't ready to merge yet)

**Comment by MTRNord** [comment]
> > Looks fine :) Feel free to press merge if you are ready to merge this. (Not sure if this is WIP or ready? :D)
> 
> I don't see any 'Merge' button... I think only you have write access to this repository so you would have to merge the PR.
> (This is ready to merge; I would just make a 'draft pull request' or add a comment if I wasn't ready to merge yet)

Ah thought you were allowed to merge things :)

**Comment by MTRNord** [comment]
Will need to see when I am able to do another release :) (If I dont do it until something like next week friday feel free to give me a ping here :) I try to not forget)

**Pull request merged**